[MRG] Fix undefined behavior in _tree.pyx and  _quad_tree.pyx

slackito opened the pull request:
Accessing a member through a null pointer is undefined behavior, even if
no actual memory access is performed (like in this case, where it was
used only for offset computation). See the following link for an
explanation:
https://software.intel.com/content/www/us/en/develop/blogs/null-pointer-dereferencing-causes-undefined-behavior.html

The current approach will also fail when building with tools like ubsan
(undefined behavior sanitizer).

The standard way to compute offsets of struct members is the offsetof
macro but it seems it's not supported by Cython so I've used the
approach described here:

https://mail.python.org/pipermail/cython-devel/2013-April/003505.html

slackito committed:
Fix undefined behavior in _tree.pyx and  _quad_tree.pyx

Accessing a member through a null pointer is undefined behavior, even if
no actual memory access is performed (like in this case, where it was
used only for offset computation). See the following link for an
explanation:
https://software.intel.com/content/www/us/en/develop/blogs/null-pointer-dereferencing-causes-undefined-behavior.html

The current approach will also fail when building with tools like ubsan
(undefined behavior sanitizer).

The standard way to compute offsets of struct members is the offsetof
macro but it seems it's not supported by Cython so I've used the
approach described here:

https://mail.python.org/pipermail/cython-devel/2013-April/003505.html

slackito commented:
Ping

thomasjpfan commented on sklearn/neighbors/_quad_tree.pyx:
```diff
@@ -39,18 +41,18 @@ CELL_DTYPE = np.dtype({
-        <Py_ssize_t> &(<Cell*> NULL).barycenter,
-        <Py_ssize_t> &(<Cell*> NULL).min_bounds,
-        <Py_ssize_t> &(<Cell*> NULL).max_bounds,
+        (<Py_ssize_t>&(dummy.parent) - <Py_ssize_t>&(dummy)),
```
The mailing list states to do this:

```
cdef Struct tmp
cdef Py_ssize_t offset
offset = <Py_ssize_t> (<Py_intptr_t>&(tmp.field) - <Py_intptr_t>&(tmp))
```

with castinig to `Py_intptr_t` first, do you think this is better?

slackito replied:
Hi! Thanks for reviewing this.

I think some form of `intptr_t` is the right tool for the job, and I tried `Py_intptr_t` first, but I couldn't make it work because of `'Py_intptr_t' is not a type identifier` errors when I tried to build.

I haven't been able to find `Py_intptr_t` available as an import anywhere, but I've digged a little more and I've found that we can import the actual C `intptr_t` type from `libc.stdint` and it works.

Should I go with that instead? Do you have any advice about where to get `Py_intptr_t` from?

thomasjpfan replied:
I think the current implementation is okay. It could be that `Py_ssize_t` is typedefed to `Py_intptr_t` anyways:

https://github.com/cython/cython/blob/0a25de5b1023fd551f5e600583de8d8c87b7796d/Cython/Utility/MemoryView.pyx#L37

slackito replied:
Sorry for the github newbie question: if you think the current implementation is okay should I mark this conversation as resolved?

thomasjpfan replied:
I personally do not resolve them on my PRs, because resolved conversations are automatically hidden and could be interesting to other reviewers.

That being said, you can mark the conversation as resolved.

slackito replied:
Thanks for the explanation. I'll leave it open then :)

thomasjpfan reviewed (approved):
LGTM

rth commented:
Isn't there a way to get rid of this whole `CELL_DTYPE` dict that was originally there for old numpy as far as I understand? Similar to https://github.com/scikit-learn/scikit-learn/pull/15097/files#diff-b071106a87a03ee4e9149e3f0a2b1180L187 where it might be equivalent to something along the lines of,
```cython
cdef Cell dummy;
CELL_DTYPE = np.asarray(<Cell[:1]>(&dummy)).dtype
```
though I haven't looked at how this works in detail.

slackito commented:
> Isn't there a way to get rid of this whole `CELL_DTYPE` dict that was originally there for old numpy as far as I understand? Similar to https://github.com/scikit-learn/scikit-learn/pull/15097/files#diff-b071106a87a03ee4e9149e3f0a2b1180L187 

I'll give it a try and report back. I'm a total newbie at pretty much everything around this project, so I won't be able to go much farther than "I made that change and tests pass/don't pass", but we should get a little bit of useful signal out of that at least.

slackito commented:
I tried the approach suggested by @rth and it looks like it builds and tests in the affected directories still pass. If nobody objects I'll modify this pull request tomorrow to use this approach, which is way simpler :)

slackito committed:
Simplify definitions of types for numpy.

According to
https://github.com/scikit-learn/scikit-learn/pull/17228/commits/9e9cdb0192394d3797f4a1adc8c28db7ee05acf2#issuecomment-638998985
these dictionaries with offsets are there for old numpy versions and
they are no longer needed. We can use instead an approach similar to
https://github.com/scikit-learn/scikit-learn/pull/15097/files#diff-b071106a87a03ee4e9149e3f0a2b1180L187

I have verified that tests keep passing and that the undefined behavior
that motivated this pull request in the first place is still fixed.

slackito commented:
Any other ideas?

NicolasHug commented:
If we declare ` CELL_DTYPE = np.asarray(<Cell[:1]>(&dummy)).dtype `, can we still access the fields of a variable? I.e. can we still do `some_cell['parent']` from Python?

Also, it seems that the original concern can be addressed by changing 
`<Py_ssize_t> &(<Cell*> NULL).parent, `

into 

`<Py_ssize_t> &(<Cell*> &dummy).parent`?

I'm not a fan of `CELL_DTYPE = np.asarray(<Cell[:1]>(&dummy)).dtype` because it's very cryptic.

> In particular I don't understand why it's necessary to do the [:1] slice

no idea either, I'm even surprised the syntax is correct

slackito commented:
As I understand it, the intent of `<Py_ssize_t> &(<Cell*> NULL).parent` was to get the offset of `parent` inside a `Cell`. This works if the null pointer is address 0 because then the address of the field is equal to its offset inside the object.

Would `<Py_ssize_t> &(<Cell*> &dummy).parent` accomplish the same thing? It seems to me it would return the address of the `parent` field in the `dummy` object, and not the offset (which is why my original commit and the reference link in the description subtract the address of the `dummy` object itself to get the relative offset, i.e. `(<Py_ssize_t>&(dummy.parent) - <Py_ssize_t>&(dummy))`).

As for your other questions, sadly I have no idea. I'm willing to try any reasonable suggestions.

NicolasHug commented:
> It seems to me it would return the address of the parent field in the dummy object, and not the offset

Yes you're probably right

Can we declare the dtype with a list instead of a dict, i.e.:

```py

CELL_DTYPE = np.dtype([
	('parent', np.intp),
	('children', np.intp[8]),  # not sure about this syntax??
	...
	],
	aligned=True
)
```

the `aligned=True` should take care of the offsets

rth commented:
Asked a [SO question](https://stackoverflow.com/questions/62448946/creating-a-numpy-datatype-from-cython-struct) about it.

> Can we declare the dtype with a list instead of a dict, i.e.:

Well it we can make it work without manually defining dtypes that need to be in sync with the struct that would be preferrable. We already have such code in binary trees we just need to understand what it does. Let's see if anyone replied, otherwise I could reach out on some numpy channel.

NicolasHug commented:
OK let's see if we can get an explanation.

Though IMHO the fact that none of us here really understand what it does is an indication that we shouldn't be using it (the cycle is going to repeat once we're not around anymore). Repeating the dtype is verbose but it's not a big  deal since these almost never change. FWIW, that's what we're doing in `sklearn/ensemble/_hist_gradient_boosting/common.pyx-pxd`

rth commented:
So we got an answer in https://stackoverflow.com/a/62449233/1791279 . I am continuously impressed by the knowledge of SO participants on sometimes obscure topics.

> Repeating the dtype is verbose but it's not a big deal since these almost never change. 

Yes, but do we know what happens if one side changes and not the other. Repeating might be OK, personally I'm not too comfortable with the pointer arithmetic initially done in this PR. I think using this solution with a link to the SO comment might be enough?

Actually it looks like we merged a similar change earlier in https://github.com/scikit-learn/scikit-learn/pull/16141 . The original motivation was a comment by Jake that that was a cleaner way to do it https://github.com/scikit-learn/scikit-learn/blame/c45721d538d36ab4c322d18e33d9c10b55f5fe27/sklearn/neighbors/_binary_tree.pxi#L184

any opinions @jnothman @jeremiedbb ?

rth commented:
Also we got this original issue by manually setting offsets , so if we can rely on numpy more to do it, it would more reliable even it's a more black box solution. As long as it works.

NicolasHug commented:
> do we know what happens if one side changes and not the other

From what I tried, either a segfault or a ValueError like
`ValueError: Buffer dtype mismatch, expected 'unsigned char' but got end in 'node_struct.blah'`

jeremiedbb commented:
`<Cell[:1]>(&dummy)` is not slicing. When you cast in cython you have to provide the length of the array so it's just cast dummy to an array of Cell of length 1.

I can't find the documentation for it though. [here it is](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#coercion-to-numpy)

I think jake's comment is right. It's cleaner and it's actually the recommended way to do that. Maybe we can add a small comment to explain that to avoid future confusion.

rth reviewed (dismissed):
I'm OK with the current solution then. Thanks @slackito !

NicolasHug commented:
if we merge this we should document it

NicolasHug commented:
> If we declare CELL_DTYPE = np.asarray(<Cell[:1]>(&dummy)).dtype, can we still access the fields of a variable? I.e. can we still do some_cell['parent'] from Python?

Surprisingly this still works

slackito committed:
Document the new numpy type definition approach.

Added a brief comment and a link to stackoverflow thread that explain
why the simpler approach to define numpy types works.

slackito commented:
Thanks everyone for the comments!

Added comments with a brief explanation an a link to the stackoverflow thread posted above. Please take another look.

adrinjalali commented on sklearn/neighbors/_quad_tree.pyx:
```diff
@@ -29,30 +29,12 @@ cdef extern from "numpy/arrayobject.h":
-        <Py_ssize_t> &(<Cell*> NULL).max_bounds,
-    ]
-})
+# Build the corresponding numpy dtyle for Cell.
```
```suggestion
# Build the corresponding numpy dtype for Cell.
```

slackito replied:
Fixed, thanks!

adrinjalali commented on sklearn/tree/_tree.pyx:
```diff
@@ -67,22 +67,12 @@ cdef SIZE_t _TREE_LEAF = TREE_LEAF
-        <Py_ssize_t> &(<Node*> NULL).weighted_n_node_samples
-    ]
-})
+# Build the corresponding numpy dtyle for Node.
```
same here

slackito replied:
Fixed, thanks!

adrinjalali reviewed (commented):
Would we have a better place to document this? Is the inline comment enough? I really don't know.

slackito committed:
Fix typo 'dtyle'->'dtype'

rth reviewed (approved):
Thanks @slackito !
